base/compiler/tfuncs.jl

```diff
@@ -653,6 +678,43 @@ add_tfunc(getfield, 2, 3, getfield_tfunc, 1)
 add_tfunc(setfield!, 3, 3, (@nospecialize(o), @nospecialize(f), @nospecialize(v)) -> v, 3)
 fieldtype_tfunc(@nospecialize(s0), @nospecialize(name), @nospecialize(inbounds)) =
     fieldtype_tfunc(s0, name)
+
+function fieldtype_nothrow(@nospecialize(s0), @nospecialize(name))
+    if s0 === Any || s0 === Type || DataType ⊑ s0 || UnionAll ⊑ s0
+        # We have no idea
+        return false
+    end
+
+    if !isa(name, Const) || (!isa(name.val, Symbol) || !isa(name.val, Int))
+        # Due to bounds checking, we can't say anything unless we know what
+        # the name is.
+        return false
+    end
+
+    s = instanceof_tfunc(s0)[1]
+    if isa(s, Union)
+        return _fieldtype_nothrow(s.a, name) && _fieldtype_nothrow(s.b, name)
+    end
+    return _fieldtype_nothrow(s, name)
+end
+
+function _fieldtype_nothrow(@nospecialize(s), @nospecialize(name))
+    fld = name.val
```

Comment:
`name` can't be both `nospecialize` and `.val`. Seems like this is known to be `name::Const`?